sources/awskinesis/awskinesis.go

```diff
@@ -69,12 +82,56 @@ func (a AWSKinesis) validate() error {
 }
 
 // Fetch retrieves the next document from the awskinesis source
-func (a AWSKinesis) Fetch() ([]byte, error) {
-	// getrecord here
-	return nil, nil
+func (a AWSKinesis) Fetch(shardID uint) error {
+	// set up the shard iterator for our particular shardID
+	// NOTE: may want to make the ShardIteratorType into a config value
+	//       https://docs.aws.amazon.com/sdk-for-go/api/service/kinesis/#GetShardIteratorInput
+	iter, err := a.Service.GetShardIterator(
+		&kinesis.GetShardIteratorInput{
+			ShardId:           a.Shards[shardID].ShardId,
+			ShardIteratorType: aws.String("TRIM_HORIZON"),
+			StreamName:        aws.String(a.StreamName),
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	for {
+		records, err := a.Service.GetRecords(&kinesis.GetRecordsInput{
+			ShardIterator: iter.ShardIterator,
+		})
+		if err != nil {
+			return err
+		}
+
+		for _, rec := range records.Records {
+			if err := a.sendToEventGateway(rec); err != nil {
+				return err
+			}
+		}
+
+		if isShardClosed(records.NextShardIterator, iter.ShardIterator) {
+			return nil
+		}
+		iter.ShardIterator = records.NextShardIterator
+	}
+}
+
+// isShardClosed checks to make sure the next iterator in the record set is not nil, otherwise we're
+// at the end of our record stream.
+// https://github.com/harlow/kinesis-consumer/blob/master/consumer.go#L209
+func isShardClosed(next, curr *string) bool {
+	return next == nil || curr == next
+}
+
+// sendToEventGateway sends the data over to our event gateway
+func (a AWSKinesis) sendToEventGateway(r *kinesis.Record) error {
```

Comment:
This should probably not be implemented here but in the worker pool. But it's fine for PoC

Reply:
Concur, this should be a centralized function. Will fix.